pandas/_libs/hashtable_class_helper.pxi.in

```diff
@@ -1179,6 +1179,8 @@ cdef class StringHashTable(HashTable):
         use_na_value = na_value is not None
 
         # assign pointers and pre-filter out missing (if ignore_na)
+        # https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#caveats-when-using-a-python-string-in-a-c-context
+        keep_bad_unicode_refs = []
```

Comment:
This doesn't fix the underlying issue. It may make it _less_ likely but the lifecycle is still not properly controlled

I think what needs to be done is something akin to:

```python
'1 \udcd6a NY'.encode('utf8', errors="surrogatepass").decode('utf8', errors="surrogatepass")
```

I think you can just do that directly in Cython, but we can look at the underlying C API if needed

(side note - I would really love to get rid of `get_c_string` - not a value added layer of indirection)

Reply:
Oh OK I see the difference - CPython can cache the UTF8 bytes of a string alongside the string object. When the string is not utf8 encodable and we have to create temporary objects we run into trouble. So the list here artificially extends the lifetime of new objects guaranteed to be utf8 encodable so we can rely on that caching mechanism

To be honest I would be -1 on this change and would rather call it a wontfix. It is a very niche issue that fights the internals of CPython (and for that matter pyarrow, whose strings are utf8).

If someone wanted mixed-encoding Python strings like this I think pa.binary() is a better data type choice

Reply:
OK I'll close this PR then. My main motivation was to make `test_unique_bad_unicode` not flaky due to this issue, but I'll open a follow up PR making this test permanently`xfail(strict=True)`instead

Reply:
I think the name _test_unique_bad_unicode_ is a misnomer. The code point "\ud83d" exists in Unicode. It is a high surrogate

https://www.unicode.org/charts/PDF/UD800.pdf

The problem is that by itself that high surrogate doesn't mean anything (it would need to be paired with a low surrogate). As such, it doesn't represent any glyph in any encoding. 

AFAIU if you wanted to keep that unicode code point, you would have to:

1. Convert the Python Unicode object to bytes via `str.encode(<encoding>, errors="surrogatepass")`
2. Run your algorithms against the surrogatepass bytes
3. Convert your surrogatepassed bytes back to a unicode string via `bytes.decode(<encoding>, errors="surrogatepass")

I realize step 2 may not exist today but is a good impetus to work on interop with pa.binary() if this is required